File: lib/ai.php
<?php
/**
 * inphub: provider-agnostic AI layer (Claude, Ollama, or LM Studio).
 *
 * One entry point, ai_generate(), that feature code calls without caring which
 * provider is configured. Everything reads the current user's settings. AI is
 * optional and off by default; ai_available() gates every feature.
 */

declare(strict_types=1);

require_once __DIR__ . '/helpers.php';

/**
 * Resolve the AI config for a user from settings.
 *
 * @return array{enabled:bool, provider:string, claude_api_key:string,
 *               claude_model:string, ollama_base_url:string, ollama_model:string,
 *               lmstudio_base_url:string, lmstudio_model:string, lmstudio_api_key:string}
 */
function ai_config(int $userId): array
{
    $s = all_settings($userId);
    return [
        'enabled'         => ($s['ai_enabled'] ?? '0') === '1',
        'provider'        => $s['ai_provider'] ?? 'claude',
        'claude_api_key'  => $s['claude_api_key'] ?? '',
        'claude_model'    => $s['claude_model'] ?? 'claude-sonnet-5',
        'ollama_base_url' => $s['ollama_base_url'] ?? 'http://127.0.0.1:11434',
        'ollama_model'    => $s['ollama_model'] ?? 'llama3.1',
        'lmstudio_base_url' => $s['lmstudio_base_url'] ?? 'http://127.0.0.1:1234',
        'lmstudio_model'    => $s['lmstudio_model'] ?? '',
        'lmstudio_api_key'  => $s['lmstudio_api_key'] ?? '',
    ];
}

/** The model the configured provider will use, the one place provider maps to a model key. */
function ai_active_model(array $c): string
{
    switch ($c['provider']) {
        case 'ollama':   return $c['ollama_model'];
        case 'lmstudio': return $c['lmstudio_model'];
        default:         return $c['claude_model'];
    }
}

/**
 * True only if AI is enabled AND the selected provider is configured enough
 * to make a call.
 */
function ai_available(int $userId): bool
{
    $c = ai_config($userId);
    if (!$c['enabled']) {
        return false;
    }
    if ($c['provider'] === 'claude') {
        return $c['claude_api_key'] !== '' && $c['claude_model'] !== '';
    }
    if ($c['provider'] === 'ollama') {
        return $c['ollama_base_url'] !== '' && $c['ollama_model'] !== '';
    }
    if ($c['provider'] === 'lmstudio') {
        // The API key is optional, LM Studio only asks for one when auth is on.
        return $c['lmstudio_base_url'] !== '' && $c['lmstudio_model'] !== '';
    }
    return false;
}

/**
 * Generate a completion. Dispatches to the user's configured provider.
 *
 * @param array $opts  Optional: max_tokens (int), timeout (int seconds),
 *                     model (string, per-call override, does NOT touch settings),
 *                     raw_system (bool, skip the locale preamble; for prompts
 *                     that must come back as strict JSON).
 * @return string  The model's text reply.
 * @throws RuntimeException on any transport/API error.
 */
function ai_generate(int $userId, string $system, string $userPrompt, array $opts = []): string
{
    $c = ai_config($userId);
    $maxTokens = (int) ($opts['max_tokens'] ?? 1024);
    $timeout   = (int) ($opts['timeout'] ?? 60);

    // Every feature funnels through here, so app-wide facts the model must not
    // get wrong (currency, today's date) are injected once, in one place.
    // Prompts that demand strict JSON back opt out with raw_system.
    if (!($opts['raw_system'] ?? false)) {
        $system = ai_locale_preamble($userId) . "\n\n" . $system;
    }

    // Per-call model override (e.g. the chat's session-only selector).
    $model = isset($opts['model']) && is_string($opts['model']) ? trim($opts['model']) : '';
    if ($model !== '') {
        $c['claude_model'] = $model;
        $c['ollama_model'] = $model;
        $c['lmstudio_model'] = $model;
    }

    if ($c['provider'] === 'ollama') {
        return ai_call_ollama($c, $system, $userPrompt, $maxTokens, $timeout);
    }
    if ($c['provider'] === 'lmstudio') {
        return ai_call_lmstudio($c, $system, $userPrompt, $maxTokens, $timeout);
    }
    return ai_call_claude($c, $system, $userPrompt, $maxTokens, $timeout);
}

/**
 * App-wide facts prepended to every system prompt.
 *
 * Without this the model sees bare numbers like "2000.00" and reads them as
 * dollars, the app stores lira by default, so a 2000 TRY week was being
 * reported back as "$2000". Nothing else in the stack states the unit.
 */
function ai_locale_preamble(int $userId): string
{
    $currency = default_currency($userId);
    $name     = currency_name($currency);

    return "Context for this conversation:\n"
        . '- Today is ' . date('Y-m-d') . ' (' . date('l') . ").\n"
        . "- Every money amount in this app — and in anything you are shown below — is in "
        . "{$currency} ({$name}). Amounts are written like \"2000.00 {$currency}\".\n"
        . "- Never assume dollars, never convert an amount into another currency, and never "
        . "prefix an amount with a currency symbol other than {$currency}'s.";
}

/**
 * List models selectable for the user's configured provider.
 * Ollama: live from GET /api/tags. LM Studio: live from GET /v1/models.
 * Claude: a small curated list (the combo box is editable, so any model id can
 * still be typed).
 *
 * @return array{provider:string, default:string, models:string[]}
 */
function ai_list_models(int $userId): array
{
    $c = ai_config($userId);

    if ($c['provider'] === 'ollama') {
        $models = [];
        $base = rtrim($c['ollama_base_url'], '/');
        $ch = curl_init($base . '/api/tags');
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT        => 15,
            CURLOPT_CONNECTTIMEOUT => 10,
        ]);
        $raw = curl_exec($ch);
        curl_close($ch);
        if (is_string($raw)) {
            $body = json_decode($raw, true);
            foreach ($body['models'] ?? [] as $m) {
                if (isset($m['name']) && is_string($m['name'])) {
                    $models[] = $m['name'];
                }
            }
        }
        return ['provider' => 'ollama', 'default' => $c['ollama_model'], 'models' => $models];
    }

    if ($c['provider'] === 'lmstudio') {
        $models = [];
        $ch = curl_init(ai_lmstudio_base($c) . '/v1/models');
        curl_setopt_array($ch, [
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_HTTPHEADER     => ai_lmstudio_headers($c),
            CURLOPT_TIMEOUT        => 15,
            CURLOPT_CONNECTTIMEOUT => 10,
        ]);
        $raw = curl_exec($ch);
        curl_close($ch);
        if (is_string($raw)) {
            $body = json_decode($raw, true);
            foreach ($body['data'] ?? [] as $m) {
                // LM Studio lists embedding models alongside chat models.
                if (isset($m['id']) && is_string($m['id']) && stripos($m['id'], 'embed') === false) {
                    $models[] = $m['id'];
                }
            }
        }
        return ['provider' => 'lmstudio', 'default' => $c['lmstudio_model'], 'models' => $models];
    }

    return [
        'provider' => 'claude',
        'default'  => $c['claude_model'],
        'models'   => ['claude-fable-5', 'claude-opus-4-8', 'claude-sonnet-5', 'claude-haiku-4-5-20251001'],
    ];
}

/** Low-level Claude Messages API call. */
function ai_call_claude(array $c, string $system, string $userPrompt, int $maxTokens, int $timeout): string
{
    if ($c['claude_api_key'] === '') {
        throw new RuntimeException('Claude API key is not set.');
    }

    $payload = [
        'model'      => $c['claude_model'],
        'max_tokens' => $maxTokens,
        'system'     => $system,
        'messages'   => [
            ['role' => 'user', 'content' => $userPrompt],
        ],
    ];

    [$status, $body, $err] = ai_http_post(
        'https://api.anthropic.com/v1/messages',
        [
            'x-api-key: ' . $c['claude_api_key'],
            'anthropic-version: 2023-06-01',
            'content-type: application/json',
        ],
        $payload,
        $timeout
    );

    if ($err !== null) {
        throw new RuntimeException('Claude request failed: ' . $err);
    }
    if ($status >= 400) {
        $msg = is_array($body) && isset($body['error']['message']) ? $body['error']['message'] : 'HTTP ' . $status;
        throw new RuntimeException('Claude error: ' . $msg);
    }

    $text = $body['content'][0]['text'] ?? null;
    if (!is_string($text)) {
        throw new RuntimeException('Claude returned an unexpected response.');
    }
    return $text;
}

/** Low-level Ollama chat API call. */
function ai_call_ollama(array $c, string $system, string $userPrompt, int $maxTokens, int $timeout): string
{
    $base = rtrim($c['ollama_base_url'], '/');

    $payload = [
        'model'    => $c['ollama_model'],
        'stream'   => false,
        'messages' => [
            ['role' => 'system', 'content' => $system],
            ['role' => 'user',   'content' => $userPrompt],
        ],
        'options'  => ['num_predict' => $maxTokens],
    ];

    [$status, $body, $err] = ai_http_post(
        $base . '/api/chat',
        ['content-type: application/json'],
        $payload,
        $timeout
    );

    if ($err !== null) {
        throw new RuntimeException('Ollama request failed: ' . $err);
    }
    if ($status >= 400) {
        $msg = is_array($body) && isset($body['error']) ? $body['error'] : 'HTTP ' . $status;
        throw new RuntimeException('Ollama error: ' . $msg);
    }

    $text = $body['message']['content'] ?? null;
    if (!is_string($text)) {
        throw new RuntimeException('Ollama returned an unexpected response.');
    }
    return $text;
}

/**
 * LM Studio base URL without a trailing slash or /v1: its server tab shows
 * "http://host:1234/v1", and pasting that should work too.
 */
function ai_lmstudio_base(array $c): string
{
    return (string) preg_replace('#/v1$#', '', rtrim($c['lmstudio_base_url'], '/'));
}

/** Request headers for LM Studio; the bearer token only when one is set. */
function ai_lmstudio_headers(array $c): array
{
    $headers = ['content-type: application/json'];
    if ($c['lmstudio_api_key'] !== '') {
        $headers[] = 'Authorization: Bearer ' . $c['lmstudio_api_key'];
    }
    return $headers;
}

/** Low-level LM Studio call (OpenAI-compatible chat completions). */
function ai_call_lmstudio(array $c, string $system, string $userPrompt, int $maxTokens, int $timeout): string
{
    // No max_tokens on purpose: the model runs locally, and a cap would cut a
    // reasoning model off mid-thought before it answers. $maxTokens is kept only
    // so every provider shares one call signature.
    $payload = [
        'model'    => $c['lmstudio_model'],
        'stream'   => false,
        'messages' => [
            ['role' => 'system', 'content' => $system],
            ['role' => 'user',   'content' => $userPrompt],
        ],
    ];

    [$status, $body, $err] = ai_http_post(
        ai_lmstudio_base($c) . '/v1/chat/completions',
        ai_lmstudio_headers($c),
        $payload,
        $timeout
    );

    if ($err !== null) {
        throw new RuntimeException('LM Studio request failed: ' . $err);
    }
    if ($status >= 400) {
        $error = is_array($body) ? ($body['error'] ?? null) : null;
        $msg = is_array($error) && isset($error['message']) ? $error['message']
            : (is_string($error) ? $error : 'HTTP ' . $status);
        throw new RuntimeException('LM Studio error: ' . $msg);
    }

    $message = $body['choices'][0]['message'] ?? null;
    if (!is_array($message)) {
        throw new RuntimeException('LM Studio returned an unexpected response.');
    }
    // Only the answer comes back to the app. LM Studio either puts the thinking
    // in reasoning_content (never read here) or inlines it as a leading
    // <think> block, which is dropped because the strict-JSON prompts
    // (quick-add, repo analysis) cannot parse around it.
    $text = is_string($message['content'] ?? null) ? $message['content'] : '';
    return ltrim((string) preg_replace('#^\s*<think>.*?</think>#s', '', $text));
}

/**
 * Shared cURL POST returning [status, decodedBody, errorOrNull].
 *
 * @return array{0:int,1:mixed,2:?string}
 */
function ai_http_post(string $url, array $headers, array $payload, int $timeout): array
{
    $ch = curl_init($url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_POST           => true,
        CURLOPT_HTTPHEADER     => $headers,
        CURLOPT_POSTFIELDS     => json_encode($payload, JSON_UNESCAPED_UNICODE),
        CURLOPT_TIMEOUT        => $timeout,
        CURLOPT_CONNECTTIMEOUT => 15,
    ]);
    $raw    = curl_exec($ch);
    $status = (int) curl_getinfo($ch, CURLINFO_HTTP_CODE);
    $curlErr = curl_error($ch);
    curl_close($ch);

    if ($raw === false) {
        return [0, null, $curlErr ?: 'Connection failed'];
    }
    return [$status, json_decode($raw, true), null];
}

/**
 * Ping the configured provider to verify credentials/connectivity.
 *
 * @return array{ok:bool, error:?string, model:?string}
 */
function ai_test_connection(int $userId): array
{
    $c = ai_config($userId);
    try {
        $reply = ai_generate($userId, 'You are a connectivity test.', 'Reply with the single word: ok', ['max_tokens' => 16, 'timeout' => 30, 'raw_system' => true]);
        return ['ok' => true, 'error' => null, 'model' => ai_active_model($c)];
    } catch (Throwable $e) {
        return ['ok' => false, 'error' => $e->getMessage(), 'model' => null];
    }
}

File: server/Services/Ai.cs
using System.Globalization;
using System.Text;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using Inphub.Core;

namespace Inphub.Services;

public record AiConfig(
    bool Enabled, string Provider,
    string ClaudeApiKey, string ClaudeModel,
    string OllamaBaseUrl, string OllamaModel,
    string LmStudioBaseUrl, string LmStudioModel, string LmStudioApiKey);

// The provider-agnostic AI layer: Claude, Ollama or LM Studio, chosen in the user's settings.
public static class Ai
{
    static readonly HttpClient Http = new(new SocketsHttpHandler { ConnectTimeout = TimeSpan.FromSeconds(15) })
    {
        Timeout = Timeout.InfiniteTimeSpan,
    };

    // The user's AI settings with their defaults.
    public static AiConfig Config(int uid)
    {
        var s = Settings.All(uid);
        string Get(string key, string fallback) => s.TryGetValue(key, out var v) ? v ?? "" : fallback;
        return new AiConfig(
            Enabled: Get("ai_enabled", "0") == "1",
            Provider: Get("ai_provider", "claude"),
            ClaudeApiKey: Get("claude_api_key", ""),
            ClaudeModel: Get("claude_model", "claude-sonnet-5"),
            OllamaBaseUrl: Get("ollama_base_url", "http://127.0.0.1:11434"),
            OllamaModel: Get("ollama_model", "llama3.1"),
            LmStudioBaseUrl: Get("lmstudio_base_url", "http://127.0.0.1:1234"),
            LmStudioModel: Get("lmstudio_model", ""),
            LmStudioApiKey: Get("lmstudio_api_key", ""));
    }

    // The model the configured provider will use; the one place a provider maps to a model setting.
    public static string ActiveModel(AiConfig c) => c.Provider switch
    {
        "ollama" => c.OllamaModel,
        "lmstudio" => c.LmStudioModel,
        _ => c.ClaudeModel,
    };

    // The one gate for every AI feature: switched on and the chosen provider configured enough to call.
    public static bool Available(int uid)
    {
        var c = Config(uid);
        if (!c.Enabled) return false;
        return c.Provider switch
        {
            "claude" => c.ClaudeApiKey != "" && c.ClaudeModel != "",
            "ollama" => c.OllamaBaseUrl != "" && c.OllamaModel != "",
            "lmstudio" => c.LmStudioBaseUrl != "" && c.LmStudioModel != "",
            _ => false,
        };
    }

    // Asks the configured provider; model overrides the setting for this call only and is never saved.
    public static async Task<string> Generate(int uid, string system, string prompt,
        int maxTokens = 1024, int timeout = 60, string? model = null, bool rawSystem = false)
    {
        var c = Config(uid);
        if (!rawSystem) system = LocalePreamble(uid) + "\n\n" + system;

        var overrideModel = model?.Trim() ?? "";
        if (overrideModel != "")
        {
            c = c with { ClaudeModel = overrideModel, OllamaModel = overrideModel, LmStudioModel = overrideModel };
        }

        return c.Provider switch
        {
            "ollama" => await CallOllama(c, system, prompt, maxTokens, timeout),
            "lmstudio" => await CallLmStudio(c, system, prompt, timeout),
            _ => await CallClaude(c, system, prompt, maxTokens, timeout),
        };
    }

    // Facts every prompt gets: today's date and the currency, so "2000.00" is never read as dollars.
    public static string LocalePreamble(int uid)
    {
        var currency = Money.DefaultCurrency(uid);
        var name = Money.CurrencyName(currency);
        var now = DateTime.Now;
        return "Context for this conversation:\n"
            + $"- Today is {now:yyyy-MM-dd} ({now.ToString("dddd", CultureInfo.InvariantCulture)}).\n"
            + $"- Every money amount in this app — and in anything you are shown below — is in {currency} ({name}). "
            + $"Amounts are written like \"2000.00 {currency}\".\n"
            + $"- Never assume dollars, never convert an amount into another currency, and never prefix an amount "
            + $"with a currency symbol other than {currency}'s.";
    }

    // The models the chat's picker offers: live from Ollama / LM Studio, a short list for Claude.
    public static async Task<object> ListModels(int uid)
    {
        var c = Config(uid);
        if (c.Provider == "ollama")
        {
            var body = await GetJson(c.OllamaBaseUrl.TrimEnd('/') + "/api/tags", []);
            var models = ModelJson.Items(ModelJson.At(body, "models"))
                .Select(m => ModelJson.StringOrNull(ModelJson.At(m, "name")))
                .OfType<string>()
                .ToList();
            return new { provider = "ollama", @default = c.OllamaModel, models };
        }
        if (c.Provider == "lmstudio")
        {
            var body = await GetJson(LmStudioBase(c) + "/v1/models", LmStudioHeaders(c));
            var models = ModelJson.Items(ModelJson.At(body, "data"))
                .Select(m => ModelJson.StringOrNull(ModelJson.At(m, "id")))
                .OfType<string>()
                .Where(id => !id.Contains("embed", StringComparison.OrdinalIgnoreCase))
                .ToList();
            return new { provider = "lmstudio", @default = c.LmStudioModel, models };
        }
        return new
        {
            provider = "claude",
            @default = c.ClaudeModel,
            models = new[] { "claude-fable-5", "claude-opus-4-8", "claude-sonnet-5", "claude-haiku-4-5-20251001" },
        };
    }

    // Pings the provider; returns ok plus the model, or the error.
    public static async Task<(bool Ok, string? Error, string? Model)> TestConnection(int uid)
    {
        var c = Config(uid);
        try
        {
            await Generate(uid, "You are a connectivity test.", "Reply with the single word: ok", maxTokens: 16, timeout: 30, rawSystem: true);
            return (true, null, ActiveModel(c));
        }
        catch (Exception e)
        {
            return (false, e.Message, null);
        }
    }

    // One Claude Messages API call.
    static async Task<string> CallClaude(AiConfig c, string system, string prompt, int maxTokens, int timeout)
    {
        if (c.ClaudeApiKey == "") throw new InvalidOperationException("Claude API key is not set.");
        var payload = new JsonObject
        {
            ["model"] = c.ClaudeModel,
            ["max_tokens"] = maxTokens,
            ["system"] = system,
            ["messages"] = new JsonArray(new JsonObject { ["role"] = "user", ["content"] = prompt }),
        };
        var (status, body, error) = await PostJson("https://api.anthropic.com/v1/messages",
            [("x-api-key", c.ClaudeApiKey), ("anthropic-version", "2023-06-01")], payload, timeout);

        if (error != null) throw new InvalidOperationException("Claude request failed: " + error);
        if (status >= 400)
        {
            var message = ModelJson.At(body, "error", "message") is JsonValue m ? ModelJson.Text(m) : "HTTP " + status;
            throw new InvalidOperationException("Claude error: " + message);
        }
        return ModelJson.StringOrNull(ModelJson.At(body, "content", "0", "text"))
            ?? throw new InvalidOperationException("Claude returned an unexpected response.");
    }

    // One Ollama /api/chat call.
    static async Task<string> CallOllama(AiConfig c, string system, string prompt, int maxTokens, int timeout)
    {
        var payload = new JsonObject
        {
            ["model"] = c.OllamaModel,
            ["stream"] = false,
            ["messages"] = new JsonArray(
                new JsonObject { ["role"] = "system", ["content"] = system },
                new JsonObject { ["role"] = "user", ["content"] = prompt }),
            ["options"] = new JsonObject { ["num_predict"] = maxTokens },
        };
        var (status, body, error) = await PostJson(c.OllamaBaseUrl.TrimEnd('/') + "/api/chat", [], payload, timeout);

        if (error != null) throw new InvalidOperationException("Ollama request failed: " + error);
        if (status >= 400)
        {
            var message = ModelJson.At(body, "error") is { } e ? ModelJson.Text(e) : "HTTP " + status;
            throw new InvalidOperationException("Ollama error: " + message);
        }
        return ModelJson.StringOrNull(ModelJson.At(body, "message", "content"))
            ?? throw new InvalidOperationException("Ollama returned an unexpected response.");
    }

    // One LM Studio call (OpenAI-compatible); no max_tokens so a local reasoning model is never cut off mid-thought.
    static async Task<string> CallLmStudio(AiConfig c, string system, string prompt, int timeout)
    {
        var payload = new JsonObject
        {
            ["model"] = c.LmStudioModel,
            ["stream"] = false,
            ["messages"] = new JsonArray(
                new JsonObject { ["role"] = "system", ["content"] = system },
                new JsonObject { ["role"] = "user", ["content"] = prompt }),
        };
        var (status, body, error) = await PostJson(LmStudioBase(c) + "/v1/chat/completions", LmStudioHeaders(c), payload, timeout);

        if (error != null) throw new InvalidOperationException("LM Studio request failed: " + error);
        if (status >= 400)
        {
            var e = ModelJson.At(body, "error");
            var reason = e is JsonObject && ModelJson.At(e, "message") != null ? ModelJson.Text(ModelJson.At(e, "message"))
                : ModelJson.StringOrNull(e) ?? "HTTP " + status;
            throw new InvalidOperationException("LM Studio error: " + reason);
        }
        if (ModelJson.At(body, "choices", "0", "message") is not JsonObject message)
        {
            throw new InvalidOperationException("LM Studio returned an unexpected response.");
        }
        var content = ModelJson.StringOrNull(ModelJson.At(message, "content")) ?? "";
        return ThinkBlock.Replace(content, "", 1).TrimStart();
    }

    static readonly Regex ThinkBlock = new(@"^\s*<think>.*?</think>", RegexOptions.Singleline);

    // LM Studio's base URL without a trailing slash or /v1, since its server tab shows ".../v1".
    static string LmStudioBase(AiConfig c) => Regex.Replace(c.LmStudioBaseUrl.TrimEnd('/'), "/v1$", "");

    // The bearer header, only when an LM Studio key is set.
    static (string, string)[] LmStudioHeaders(AiConfig c) =>
        c.LmStudioApiKey != "" ? [("Authorization", "Bearer " + c.LmStudioApiKey)] : [];

    // POSTs JSON; returns the status, the parsed body (or null) and a transport error (or null).
    static async Task<(int Status, JsonNode? Body, string? Error)> PostJson(string url, (string Name, string Value)[] headers,
        JsonObject payload, int timeout)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new StringContent(ModelJson.ToText(payload), Encoding.UTF8, "application/json"),
        };
        foreach (var (name, value) in headers) request.Headers.TryAddWithoutValidation(name, value);
        return await Send(request, timeout);
    }

    // GETs JSON, returning the parsed body or null on any failure.
    static async Task<JsonNode?> GetJson(string url, (string Name, string Value)[] headers)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        foreach (var (name, value) in headers) request.Headers.TryAddWithoutValidation(name, value);
        var (_, body, _) = await Send(request, 15);
        return body;
    }

    // Sends a request with a timeout in seconds; never throws, a failure comes back as the error text.
    static async Task<(int Status, JsonNode? Body, string? Error)> Send(HttpRequestMessage request, int timeout)
    {
        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(timeout));
        try
        {
            using var response = await Http.SendAsync(request, cts.Token);
            var raw = await response.Content.ReadAsStringAsync(cts.Token);
            return ((int)response.StatusCode, ModelJson.Parse(raw), null);
        }
        catch (OperationCanceledException)
        {
            return (0, null, $"Operation timed out after {timeout} seconds");
        }
        catch (HttpRequestException e)
        {
            return (0, null, e.Message);
        }
    }
}
